Extend _sharded_tensor constructor to support other ops like torch.ones

bowangbj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63378

a) Introduce InitCommonParams to wrap tensor creation params
b) Factor local tensor initiation into common_params so that tensor value is not hard specified in ShardedTensor constructor
c) Add _sharded_tensor.ones(...) to exemplify - Note memory_format arg is not provided to be consistent as torch.ones
d) Follow up: more ops like torch.full, torch.zero, torch.rand,

Test:
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestCreateTensorFromParams --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorChunked.test_create_sharded_tensor_with_ones --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorEnumerable.test_create_sharded_tensor_with_ones --v

Differential Revision: [D30359245](https://our.internmc.facebook.com/intern/diff/D30359245)

bowangbj committed:
DRAFT: sharded_tensor.ones

[ghstack-poisoned]

bowangbj committed:
Update on "DRAFT: sharded_tensor.ones"




[ghstack-poisoned]

bowangbj committed:
_sharded_tensor.ones ready for code review on "Extend _sharded_tensor constructor to support other ops like torch.ones"


a) Introduce InitCommonParams to wrap tensor creation params
b) Factor local tensor initiation into common_params so that tensor value is not hard specified in ShardedTensor constructor
c) Add _sharded_tensor.ones(...) to exemplify
d) Follow up: more ops like torch.full, torch.zero, torch.rand,

[ghstack-poisoned]

bowangbj committed:
update test on "Extend _sharded_tensor constructor to support other ops like torch.ones"


a) Introduce InitCommonParams to wrap tensor creation params
b) Factor local tensor initiation into common_params so that tensor value is not hard specified in ShardedTensor constructor
c) Add _sharded_tensor.ones(...) to exemplify - Note memory_format arg is not provided to be consistent as torch.ones
d) Follow up: more ops like torch.full, torch.zero, torch.rand,

Test:
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestCreateTensorFromParams --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorChunked.test_create_sharded_tensor_with_ones --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorEnumerable.test_create_sharded_tensor_with_ones --v

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30359245).

bowangbj committed:
fix lint on "Extend _sharded_tensor constructor to support other ops like torch.ones"


a) Introduce InitCommonParams to wrap tensor creation params
b) Factor local tensor initiation into common_params so that tensor value is not hard specified in ShardedTensor constructor
c) Add _sharded_tensor.ones(...) to exemplify - Note memory_format arg is not provided to be consistent as torch.ones
d) Follow up: more ops like torch.full, torch.zero, torch.rand,

Test:
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestCreateTensorFromParams --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorChunked.test_create_sharded_tensor_with_ones --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorEnumerable.test_create_sharded_tensor_with_ones --v

Differential Revision: [D30359245](https://our.internmc.facebook.com/intern/diff/D30359245)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30359245).

rohan-varma commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -219,6 +257,36 @@ def test_complete_world_size(self):
+                "rank:3/cuda:3",
+            ],
+        )
+        sharded_tensor = _sharded_tensor.ones(spec, 10, 20)
```
nit: store the shape in a variable and check the shards by dividing/modding for the appropriate rank? 

i.e. 
```
h, w = 10, 20
tensor = sharded.ones(spec, h, w)
...
self.assertEqual(local_shard, ones(ceil(h//world_size, w)

```

bowangbj replied:
Good point, updated per suggested

rohan-varma commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -170,25 +163,21 @@ def __init__(
-        requires_grad=False,
-        pin_memory=False,
-        memory_format=torch.contiguous_format,
+        common_params: InitCommonParams,
```
is `_internals` namespace the right place for `InitCommonParams` given that its actually a public class to the user?

bowangbj replied:
Good point, this seems an interesting discussion. From existing code, I found: 
1. there was _internals.py under _sharding_spec 
2.the dir name _sharded_tensor and _sharding_spec suggested the entire folders are still internal 
3. the actual API interfaces seems living inside _sharded_tensor.__init__ (since the main APIs we wanted to expose is _sharded_tensor.empty, .ones, .init_from_local_sharded). the _sharded_tensor/api.py seems impl details.

How about we 
a) move _sharded_tensor/internals.py to _sharded_tensor/api.py 
and b) rename _sharded_tensor/api.py to sth like _sharded_tensor/_api.py or api_impl.py or sth else?

@rohan-varma @pritamdamania

pritamdamania87 replied:
I kept the entire directory to be `_sharded_tensor` and `_sharding_spec` just as an indication that this is still initial prototype/beta work. Eventually we will rename these folders to be `sharding_spec` and `sharded_tensor` for an official release.

I think we can keep `InitCommonParams` in `api.py` itself but keep methods like `create_tensor_from_params` in internals.py

bowangbj replied:
After a 2nd thought, seems I need to remove _internals.py if we want to keep InitCommonParams (TensorInitParams) in api.py and create_tensor_from_params since that introduced circular dependency.

bowangbj committed:
Update on "Extend _sharded_tensor constructor to support other ops like torch.ones"


a) Introduce InitCommonParams to wrap tensor creation params
b) Factor local tensor initiation into common_params so that tensor value is not hard specified in ShardedTensor constructor
c) Add _sharded_tensor.ones(...) to exemplify - Note memory_format arg is not provided to be consistent as torch.ones
d) Follow up: more ops like torch.full, torch.zero, torch.rand,

Test:
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestCreateTensorFromParams --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorChunked.test_create_sharded_tensor_with_ones --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorEnumerable.test_create_sharded_tensor_with_ones --v

Differential Revision: [D30359245](https://our.internmc.facebook.com/intern/diff/D30359245)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30359245).

bowangbj committed:
resolve rohan comment on "Extend _sharded_tensor constructor to support other ops like torch.ones"


a) Introduce InitCommonParams to wrap tensor creation params
b) Factor local tensor initiation into common_params so that tensor value is not hard specified in ShardedTensor constructor
c) Add _sharded_tensor.ones(...) to exemplify - Note memory_format arg is not provided to be consistent as torch.ones
d) Follow up: more ops like torch.full, torch.zero, torch.rand,

Test:
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestCreateTensorFromParams --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorChunked.test_create_sharded_tensor_with_ones --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorEnumerable.test_create_sharded_tensor_with_ones --v

Differential Revision: [D30359245](https://our.internmc.facebook.com/intern/diff/D30359245)

[ghstack-poisoned]

bowangbj commented:
Thanks Rohan for the super quick review, resolved the test comment, added some context for the _internal.py change. PTAL

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30359245).

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -146,20 +152,7 @@ class ShardedTensor(object):
-            the default process group will be used. If specified the ShardedTensor is only
-            built on ranks that are part of this process group and the provided ``sharding_spec``
-            is applied in the context of this process group.
+        common_params (:class: `InitCommonParams`): common params to create tensor.
```
nit: Maybe just call these `tensor_init_params` and `TensorInitParams` since they basically control how the underlying tensors are initialized. The "common" part is mostly for our internal developer use which makes this simpler so its better if the user doesn't know about it.

bowangbj replied:
Great point, that made it much more readable.

pritamdamania87 reviewed (commented):
Looks great! Just have a few comments regarding naming.

bowangbj committed:
resolve pritam comment on "Extend _sharded_tensor constructor to support other ops like torch.ones"


a) Introduce InitCommonParams to wrap tensor creation params
b) Factor local tensor initiation into common_params so that tensor value is not hard specified in ShardedTensor constructor
c) Add _sharded_tensor.ones(...) to exemplify - Note memory_format arg is not provided to be consistent as torch.ones
d) Follow up: more ops like torch.full, torch.zero, torch.rand,

Test:
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestCreateTensorFromParams --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorChunked.test_create_sharded_tensor_with_ones --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorEnumerable.test_create_sharded_tensor_with_ones --v

Differential Revision: [D30359245](https://our.internmc.facebook.com/intern/diff/D30359245)

[ghstack-poisoned]

bowangbj committed:
lint on "Extend _sharded_tensor constructor to support other ops like torch.ones"


a) Introduce InitCommonParams to wrap tensor creation params
b) Factor local tensor initiation into common_params so that tensor value is not hard specified in ShardedTensor constructor
c) Add _sharded_tensor.ones(...) to exemplify - Note memory_format arg is not provided to be consistent as torch.ones
d) Follow up: more ops like torch.full, torch.zero, torch.rand,

Test:
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestCreateTensorFromParams --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorChunked.test_create_sharded_tensor_with_ones --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorEnumerable.test_create_sharded_tensor_with_ones --v

Differential Revision: [D30359245](https://our.internmc.facebook.com/intern/diff/D30359245)

[ghstack-poisoned]

bowangbj commented:
Thanks Pritam for the great comments. Renamed the param to TensorInitParams as suggested. The separation of _internals.py (move TensorInitParams to api and keep create_tensor_from_params in _internals.py) can't be done since that introduces circular dependency, b/c of that, I simply merge _internals.py into apy.py as Rohan suggested. 

PTAL

Small tips (just in case), to check what's the delta compared to last revision, you can
- Go to Files changed
- Changes from all commits - drop down
- Screen down and click latest commits [resolve pritam comment] like [this one](https://github.com/pytorch/pytorch/pull/63378/commits/5be25de94e9295a3943111360917f8a7f87cf357)

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30359245).

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -615,3 +633,26 @@ def __setstate__(self, state):
         self._post_init()
+
+
+def create_tensor_from_params(*size, local_device, tensor_init_params: TensorInitParams):
```
nit: make this `_create_tensor_from_params` since it is an internal method and isn't meant to be a public API.

bowangbj replied:
Oh yes, fixed.

bowangbj committed:
make _create_tensor_from_params internal on "Extend _sharded_tensor constructor to support other ops like torch.ones"


a) Introduce InitCommonParams to wrap tensor creation params
b) Factor local tensor initiation into common_params so that tensor value is not hard specified in ShardedTensor constructor
c) Add _sharded_tensor.ones(...) to exemplify - Note memory_format arg is not provided to be consistent as torch.ones
d) Follow up: more ops like torch.full, torch.zero, torch.rand,

Test:
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestCreateTensorFromParams --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorChunked.test_create_sharded_tensor_with_ones --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorEnumerable.test_create_sharded_tensor_with_ones --v

Differential Revision: [D30359245](https://our.internmc.facebook.com/intern/diff/D30359245)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30359245).

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30359245).

bowangbj committed:
Update on "Extend _sharded_tensor constructor to support other ops like torch.ones"


a) Introduce InitCommonParams to wrap tensor creation params
b) Factor local tensor initiation into common_params so that tensor value is not hard specified in ShardedTensor constructor
c) Add _sharded_tensor.ones(...) to exemplify - Note memory_format arg is not provided to be consistent as torch.ones
d) Follow up: more ops like torch.full, torch.zero, torch.rand,

Test:
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestCreateTensorFromParams --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorChunked.test_create_sharded_tensor_with_ones --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorEnumerable.test_create_sharded_tensor_with_ones --v

Differential Revision: [D30359245](https://our.internmc.facebook.com/intern/diff/D30359245)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30359245).

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -123,6 +125,26 @@ def _register_remote_shards(sharded_tensor_id: int, rrefs: List[rpc.RRef[Shard]]
+                 'memory_format']
+
+    create_op: CreateOp
+    dtype: torch.dtype
```
nit: just a minor suggestion and can be in a follow up diff. I noticed that these fields have a big overlap with ShardedTensorMetadata, maybe we can group these tensor relevant properties together as TensorProperties (i.e. dtype, layout, requires_grad, pin_memory, memory_format), then ShardedTensorMetadata could just take that and a list of metadatas, sharded tensor constructors can just take `create_op` and tensor property object.

bowangbj replied:
Great suggestion Wanchao, will send out a follow up PR for this.

bowangbj replied:
Follow up: sent out the consolidation PR at https://github.com/pytorch/pytorch/pull/63731

bowangbj commented:
Thanks Pritam and Wanchao. Merged the PR. Wanchao, will address your cmts in a follow up PR.
